### PR TITLE
Masked auto encoder

### PR DESCRIPTION
**Pull request by guarin**

### CODE REVIEW

**Review by MalteEbner — COMMENTED** [review]
So in total there are many steps:
1. image ----cropping-----> patches
2. patches ----- tokenisation ---> patch-wise tokens
3. patch-wise tokens ------prepend class token----> encoder input tokens
4. encoder input tokens + encoder positional embedding ------ element-wise add positional embedding ---> encoder input tokens (at all indices)
5. encoder input tokens ------ utils.get_at_index(input, idx_keep) ---> encoder input tokens (only at kept indices)
6. encoder input tokens -------transformer forward pass -----> encoder output tokens 
7. encoder output tokens ----- encoder-decoder-decoupling ---> decoder input tokens (only at kept indices)
8. decoder input tokens + masked token --------un-indexing / adding masks ----> decoder input tokens (at all indices)
9. decoder input tokens ------- decoder forward pass ----> decoder output tokens (at all indices)
10. decoder output tokens (at all indices) --- utils.get_at_index(input, idx_mask) ----> decoder output tokens (at masked indices)
10. decoder output tokens (at masked indices) ------ reverse tokenise / predict pixels from token ----> decoder predictions (at masked indices)

11a. full image -- cropping -> patches (at all indices)
11b full image -- utils.get_at_index(input, idx_mask) -> patches (at masked indices)

12 patches (at masked indices, from step 11) + decoder predictions (at masked indices, from step 10) ------> loss function

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 1]
```diff
@@ -117,15 +118,25 @@
...
+    torchvision.transforms.RandomHorizontalFlip(),
+    torchvision.transforms.ToTensor(),
+    normalize_transform,
+])
+
 # No additional augmentations for the test set
 test_transforms = torchvision.transforms.Compose([
     torchvision.transforms.Resize(input_size),
-    torchvision.transforms.CenterCrop(128),
+    torchvision.transforms.CenterCrop(224),
```
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:137]: Same as above.

**Comment by guarin** [comment]
First results on the imagenette benchmark:

```
------------------------------------------------------------------------------------------
| Model         | Batch Size | Epochs |  KNN Test Accuracy |       Time | Peak GPU Usage |
------------------------------------------------------------------------------------------
| MAE           |        256 |    200 |              0.488 |   64.3 Min |      4.6 GByte |
| MAE           |        256 |    800 |              0.681 |  272.6 Min |      4.6 GByte |
| SimCLR        |        256 |    200 |              0.771 |   82.2 Min |      3.9 GByte |
| SimCLR        |        256 |    800 |              0.858 |  324.8 Min |      3.9 GByte |
```

The performance is quite a bit lower than SimCLR. Main differences to the paper/reference implementation:
* No warmup during training
* Learned positional embedding instead of fixed one (we just reuse the positional embedding from torchvision)
* Decoder is less deep
* Reconstruction target uses non-normalized pixel values instead of normalized ones

Looking at the ablation studies from the paper the decoder depth should have the largest impact. But I doubt it will improve accuracy by 20-30% to match SimCLR accuracy.

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 2]
```diff
@@ -117,15 +118,25 @@
...
 )
 
+normalize_transform = torchvision.transforms.Normalize(
+    mean=lightly.data.collate.imagenet_normalize['mean'],
+    std=lightly.data.collate.imagenet_normalize['std'],
+)
+
+# MAE transforms
+mae_train_transforms = torchvision.transforms.Compose([
+    torchvision.transforms.RandomResizedCrop(224, scale=(0.2, 1.0), interpolation=3),  # 3 is bicubic
```
[the remark this answers is not in the document]
Comment (MalteEbner) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:129]: In this case, I would use a constant for I, something like `VIT_INPUT_SIZE`

[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py]: No longer necessary as it is handled by the MAECollateFunction

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 3]
```diff
@@ -154,6 +165,9 @@ def get_data_loaders(batch_size: int, model):
         col_fn = swav_collate_fn
     elif model == DINOModel:
         col_fn = dino_collate_fn
+    elif model == MAEModel:
```
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:172]: Added a MAECollateFunction

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 4]
```diff
@@ -577,15 +591,95 @@ def configure_optimizers(self):
...
+
+        # decoder forward pass
+        x_decoded = self.decoder.decode(x_masked)
+
+        # predict pixel values for masked tokens
+        x_pred = utils.get_at_index(x_decoded, idx_mask)
+        x_pred = self.decoder.predict(x_pred)
+        return x_pred
+
+    def training_step(self, batch, batch_idx):
```
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:724]: The length of the variables is always the batch size. If you convert all `len(x)` to `x.shape[1]` then your assumptions are correct :) I didn't add comments because it is only the benchmark module. The shapes are now well documented in the model/function docstrings. We could also add some more comments once we add the model examples.

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 5]
```diff
@@ -659,12 +668,99 @@ def configure_optimizers(self):
...
+            embed_input_dim=vit.hidden_dim,
+            hidden_dim=decoder_dim,
+            mlp_dim=decoder_dim * 4,
+            out_dim=vit.patch_size ** 2 * 3,
+            dropout=0,
+            attention_dropout=0,
+        )
+        self.criterion = nn.MSELoss()
+
+    def random_mask(self, images):
```
Comment (philippmwirth) [inline docs/source/getting_started/benchmarks/imagenette_benchmark.py:697]: This is a bit unnecessary to put it into its own function don't you think? Takes away from the "simple-to-use"

**Review by philippmwirth — APPROVED** [review]
This looks good! Did you run the benchmarks again after switching to the `CollateFunction`?

Also I'd suggest to create a follow-up issue to add it to the examples otherwise we can't really promote it nicely 🙂

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py (hunk 5 above)
Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py]: I agree, changed it :)

**Comment by guarin** [comment]
> This looks good! Did you run the benchmarks again after switching to the CollateFunction?

Not yet, I'll run it in the background.

> Also I'd suggest to create a follow-up issue to add it to the examples otherwise we can't really promote it nicely 🙂

👍🏼

**Comment by guarin** [comment]
The tests failed because of codecov upload failing. All unit tests passed.
